feat(performance): allow creating a faster capsule

imsnif opened the pull request:
## Proposed Changes
This integrates the librarian package manager to bit when building components. It allows an environment to:
1. pass a `skipNodeModules` argument when isolating, so that the capsule will know not to install with npm/yarn.
2. use `capsule.execNode` to run an executable through librarian, which patches the executables `fs` module with the required node_modules.
3. allows the environment to pass a `keepExistingCapsule` flag to bit so that the capsule will not be deleted when it is done (this is so that the next librarian installation will be cached and much faster - it can also help an environment rely on incremental builds).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/teambit/bit/2162)
<!-- Reviewable:end -->

GiladShoham commented:
@imsnif please configure the circle to be able to install the package so we can see that nothing breaks

GiladShoham commented on src/constants.ts:
```diff
@@ -4,6 +4,7 @@ import * as path from 'path';
 import format from 'string-format';
 import { PathOsBased } from './utils/path';
 import { ComponentOrigin } from './consumer/bit-map/component-map';
+process.env.MEMFS_DONT_WARN = 'true'; // suppress fs experimental warnings from memfs
```
i'm not sure this is the correct place to put it. 
maybe in the main file (bin/bit.js)?

davidfirst replied:
I'd put it in `src/app.ts`.

GiladShoham commented on src/environment/isolator.ts:
```diff
@@ -103,29 +119,31 @@ export default class Isolator {
     await this._addComponentsToRoot();
     logger.debug('ManyComponentsWriter, install packages on capsule');
-    await this._installWithPeerOption();
+    if (!opts.skipNodeModules) {
```
looks like it's the same as the installNpmPackages option. (and it's just a bug it wasn't used here).
if so, let's use the existing one?

imsnif commented:
Hey, so this is ready for another review round. At the time of this writing the AppVeyor tests are still running, but the e2e circle tests are passing. I'll address the AppVeyor errors if/when they come.

Things I did:
1. I found a bug where components with dependencies would not be updated if their dependencies changed. This involved some changes in bit (namely a smarter overriding of existing files inside a long lived capsule), as well as in librarian (already pushed to the branch) and in the typescript compiler (passing a proper flag to the isolate function and not doing an incremental typescript build, because tsc does not handle this situation well).
2. Used the installNpmPackages flag instead of the skipNodeModules flag (which I removed)
  * This also involved adjusting the isolate command to accept a `skip-npm-installation` instead of a `install-packages` flag, as well as adjusting some e2e tests to match this behaviour.
3. Moved the environment variable as per @davidfirst's recommendation
4. Fixed lint issues
5. Merged master to here

GiladShoham commented on src/environment/isolator.ts:
```diff
@@ -83,6 +98,8 @@ export default class Isolator {
     }
     const writeToPath = opts.writeToPath;
+    const installNpmPackages = typeof opts.installNpmPackages === 'undefined' ? true : opts.installNpmPackages;
+    // default should be true
```
please move the comment to be above the code line, it's above the code across the whole repo. (relevant for other comments as well)

GiladShoham commented on src/consumer/component/sources/data-to-persist.ts:
```diff
@@ -71,21 +71,29 @@ export default class DataToPersist {
-  async _writeFileToCapsule(capsule: Capsule, file: AbstractVinyl) {
-    if (file.override === false) {
+  async _writeFileToCapsule(capsule: Capsule, file: AbstractVinyl, opts = { overwriteExisting: false }) {
+    if (file.override === false || opts.overwriteExisting) {
```
I don't like the flow here, it's very confusing:
1. what is the different between `file.override` and `opts.overwriteExisting`? where do I get each one of them? who is stronger?
2. the condition is very weird, it's sounds something like - if (not override or override get here). like if (false or true) enter.
(I don't have in mind a better way, but let's try thinking about something more clear)
